packages/labs/ssr/src/lib/declarative-style-dedupe.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {digestForTemplateResult} from '@lit-labs/ssr-client';
+import {CSSResult, CSSResultOrNative, TemplateResult} from 'lit';
+
+/**
+ * StyleDedupe is a utility that can be optionally passed to your
+ * SSR
+ */
+export class DeclarativeStyleDedupeUtility {
+  seenStyles = new Set<string>();
+  private hasEmittedDeclaration = false;
+
+  emitCustomElementDeclaration(): string {
+    if (this.hasEmittedDeclaration) {
+      return '';
+    }
+    this.hasEmittedDeclaration = true;
+    return `
+<script>
+
+class StyleModule extends HTMLElement {
+    static styleCache = new Map();
+
+    connectedCallback() {
+        // This attribute tells us that we expect the child styles
+        // to be cached.
+        if (this.hasAttribute("need-cached-styles")) {
+            return this.parsedCallback();
+        }
+
+        const obs = new MutationObserver(() => {
+            obs.disconnect();
+            this.parsedCallback();
+        });
+        obs.observe(this.parentNode, {childList: true, subtree: true});
+    }
+
+    parsedCallback() {
+        const styleId = this.getAttribute('style-id');
+        if (this.children.length === 1) {
+            const styleEl = this.children[0];
+            StyleModule.styleCache.set(styleId, styleEl.textContent);
+            console.log('setting styleId', styleId, styleEl.textContent);
+            return;
+          } else if (this.children.length !== 0) {
+            throw new Error("Unexpected length of children in style-module");
+          }
+        const styleContents = StyleModule.styleCache.get(styleId);
+        const styleEl = document.createElement('style');
+        styleEl.textContent = styleContents;
+        this.append(styleEl);
+    }
+}
+customElements.define('style-module', StyleModule);
+</script>`;
+  }
+
+  getStyleHash(styles: CSSResultOrNative[]): string {
```

Comment:
I don't think you need a hash, unless we're trying to make this work over multiple requests. I think you should be able to use stylesheet identity on the server and a simple counter for the id.